python/pip.bzl

```diff
@@ -13,84 +13,67 @@
-    Label("//rules_python:whl.bzl.tpl"),
+  ctx.template(
+    "requirements.bzl",
+    Label("//rules_python:requirements.bzl.tpl"),
     substitutions = {
-      "%{repo}": repository_ctx.name,
-      "%{pip_args}": ", ".join(["\"%s\"" % arg for arg in repository_ctx.attr.pip_args]),
-      "%{requirements}": str(repository_ctx.attr.requirements_bzl),
-      "%{additional_buildtime_deps}": _expand_deps_to_dict(repository_ctx.attr.additional_buildtime_deps),
-      "%{additional_buildtime_env}": _expand_env_to_dict(repository_ctx.attr.additional_buildtime_env),
-      "%{additional_runtime_deps}": _expand_deps_to_dict(repository_ctx.attr.additional_runtime_deps),
-      "%{additional_build_content}": _expand_build_deps_to_dict(repository_ctx.attr.additional_build_content),
-      "%{remove_runtime_deps}": _expand_deps_to_dict(repository_ctx.attr.remove_runtime_deps),
-      "%{patch_runtime}": _expand_deps_to_dict(repository_ctx.attr.patch_runtime),
+      "%{repo}": ctx.name,
+      "%{python}": str(ctx.attr.python) if ctx.attr.python else "",
+      "%{pip_args}": ", ".join(["\"%s\"" % arg for arg in ctx.attr.pip_args]),
+      "%{additional_attributes}": ctx.attr.requirements_overrides or "{}",
     })
 
-  repository_ctx.template(
-    "update.sh",
-    Label("//rules_python:update.sh.tpl"),
-    substitutions = {
-      "%{python}": str(repository_ctx.path(repository_ctx.attr.python)) if repository_ctx.attr.python else "python",
-      "%{python_label}": str(repository_ctx.attr.python) if repository_ctx.attr.python else "",
-      "%{piptool}": str(repository_ctx.path(repository_ctx.attr._script)),
-      "%{name}": repository_ctx.attr.name,
-      "%{build_dependencies}": " ".join(['%s=%s' % (k, vv) for k,v in repository_ctx.attr.additional_buildtime_deps.items() for vv in v]),
-      "%{requirements_txt}": " ".join(["\"%s\"" % str(repository_ctx.path(f)) for f in repository_ctx.attr.requirements]),
-      "%{requirements_bzl}": str(repository_ctx.path(repository_ctx.attr.requirements_bzl)) if repository_ctx.attr.requirements_bzl else "",
-      "%{directory}": str(repository_ctx.path("")),
-      "%{pip_args}": " ".join(["\"%s\"" % arg for arg in repository_ctx.attr.pip_args]),
-    },
-    executable=True,
-  )
 
-  if repository_ctx.attr.requirements_bzl:
-    repository_ctx.symlink(repository_ctx.path(repository_ctx.attr.requirements_bzl), "requirements.bzl")
+  cmd = [
+    ctx.path(ctx.attr.python) if ctx.attr.python else "python",
+    ctx.path(ctx.attr._script),
+    "resolve",
+    "--name=%s" % ctx.attr.name,
+    "--pip-arg=--cache-dir=%s" % str(ctx.path("pip-cache")),
+  ] + [
+    "--input=%s" % str(ctx.path(f)) for f in ctx.attr.requirements
+  ] + [
+    "--pip-arg=%s" % x for x in ctx.attr.pip_args
+  ]
+
+
+  if ctx.attr.requirements_bzl:
+    cmd += [
+        "--output=%s" % str(ctx.path(ctx.attr.requirements_bzl)),
+        "--output-format=download",
+        "--directory=%s" % str(ctx.path("build-directory")),
```

Comment:
Is there any link to docs on this? Not sure how it works. I didn't know we could get the final file path in the repo from a rule.

Reply:
From https://docs.bazel.build/versions/master/skylark/lib/repository_ctx.html#path:
> If the path is relative, it will resolve relative to the repository directory.